services/api/service.go

```diff
@@ -470,29 +479,54 @@ func (api *RelayAPI) StartServer() (err error) {
 		IdleTimeout:       time.Duration(apiIdleTimeoutMs) * time.Millisecond,
 		MaxHeaderBytes:    apiMaxHeaderBytes,
 	}
-
 	err = api.srv.ListenAndServe()
 	if errors.Is(err, http.ErrServerClosed) {
 		return nil
 	}
 	return err
 }
 
-// StopServer disables sending any bids on getHeader calls, waits a few seconds to catch any remaining getPayload call, and then shuts down the webserver
+func (api *RelayAPI) IsReady() bool {
+	// If server is shutting down, return false
+	if api.srvShutdown.Load() {
+		return false
+	}
+
+	// Proposer API readiness checks
+	if api.opts.ProposerAPI {
+		knownValidatorsUpdated := api.datastore.KnownValidatorsWasUpdated.Load()
+		return knownValidatorsUpdated
+	}
+
+	// Block-builder API readiness checks
+	return true
+}
+
+// StopServer gracefully shuts down the HTTP server:
+// - Stop returning bids
+// - Set ready /readyz to negative status
+// - Wait a bit to allow removal of service from load balancer and draining of requests
 func (api *RelayAPI) StopServer() (err error) {
+	// avoid running this twice. setting srvShutdown to true makes /readyz switch to negative status
+	if wasStopping := api.srvShutdown.Swap(true); wasStopping {
+		return nil
+	}
+
+	// start server shutdown
 	api.log.Info("Stopping server...")
 
+	// stop returning bids on getHeader calls
 	if api.opts.ProposerAPI {
-		// stop sending bids
 		api.ffForceGetHeader204 = true
-		api.log.Info("Disabled sending bids, waiting a few seconds...")
+		api.log.Info("Disabled returning bids on getHeader")
+	}
 
-		// wait a few seconds, for any pending getPayload call to complete
-		time.Sleep(5 * time.Second)
+	// wait some time to get service removed from load balancer
+	api.log.Infof("Waiting %.2f seconds before shutdown...", apiShutdownWaitDuration.Seconds())
+	time.Sleep(apiShutdownWaitDuration)
 
-		// wait for any active getPayload call to finish
-		api.getPayloadCallsInFlight.Wait()
-	}
+	// wait for any active getPayload call to finish
+	api.getPayloadCallsInFlight.Wait()
 
 	// shutdown
 	return api.srv.Shutdown(context.Background())
```

Comment:
> Once Shutdown has been called on a server, it may not be reused; future calls to methods such as Serve will return ErrServerClosed.

There's an atomic bool that ensures that `StopServer` can only be called once, so this error can't be returned iiuc.